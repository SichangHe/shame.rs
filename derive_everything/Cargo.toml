[package]
name = "derive_everything"
version = "0.1.0"
edition = "2021"
description = "#[derive(Copy, Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)] everything if possible."
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
proc-macro = true

[dev-dependencies]
trybuild = { version = "1", features = ["diff"] }

[dependencies]
impls.workspace = true
quote = "1"
syn = { version = "2", features = ["full"] }
