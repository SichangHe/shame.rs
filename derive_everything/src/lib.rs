use proc_macro::TokenStream;
use quote::quote;
use syn::{parse::Nothing, parse_macro_input, Item};

#[proc_macro_attribute]
pub fn derive_everything(args: TokenStream, input: TokenStream) -> TokenStream {
    let _ = parse_macro_input!(args as Nothing);
    let item = parse_macro_input!(input as Item);

    quote! {
        #[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
        #item
    }
    .into()
}

#[proc_macro_attribute]
pub fn derive_enum_everything(args: TokenStream, input: TokenStream) -> TokenStream {
    let _ = parse_macro_input!(args as Nothing);
    let item = parse_macro_input!(input as Item);

    quote! {
        #[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
        #item
    }
    .into()
}

#[proc_macro_attribute]
pub fn derive_float_everything(args: TokenStream, input: TokenStream) -> TokenStream {
    let _ = parse_macro_input!(args as Nothing);
    let item = parse_macro_input!(input as Item);

    quote! {
        #[derive(Clone, Debug, Default, PartialEq, PartialOrd)]
        #item
    }
    .into()
}

#[proc_macro_attribute]
pub fn derive_float_enum_everything(args: TokenStream, input: TokenStream) -> TokenStream {
    let _ = parse_macro_input!(args as Nothing);
    let item = parse_macro_input!(input as Item);

    quote! {
        #[derive(Clone, Debug, PartialEq, PartialOrd)]
        #item
    }
    .into()
}
