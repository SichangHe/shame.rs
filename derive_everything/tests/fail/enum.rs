use derive_everything::derive_everything;

#[derive_everything]
enum MoreComplicated {
    Num(u64),
    Str { string: String },
    Vector(Vec<u64>),
}

fn main() {}
