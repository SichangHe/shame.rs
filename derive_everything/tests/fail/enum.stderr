error: no default declared
 --> tests/fail/enum.rs:3:1
  |
3 | #[derive_everything]
  | ^^^^^^^^^^^^^^^^^^^^
  |
  = help: make a unit variant default by placing `#[default]` above it
  = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)
